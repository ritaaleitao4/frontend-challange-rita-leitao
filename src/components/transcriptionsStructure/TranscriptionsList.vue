<template>
  <div class="list-content">
    <transition-group name="list-anim" tag="ul">
      <li v-for="item in listData" :key="item.id">
        <TranscriptionsListItem :item="item"></TranscriptionsListItem>
      </li>
    </transition-group>
    <add-row-svg class="list-content--add-row" @click="addNewData"/>
  </div>
</template>

<script>
  import { mapActions } from 'vuex'
  import TranscriptionsListItem from '@/components/transcriptionsStructure/TranscriptionsListItem.vue'
  import addRowSvg from '@/assets/images/add-row.svg'

  export default {
    name: 'TranscriptionsList',
    components: {
      TranscriptionsListItem,
      addRowSvg,
    },
    props: {
      listData: {
        required: true,
        type: Array,
      },
    },
    methods: {
      ...mapActions([
        'addNewTranscription',
      ]),
      addNewData() {
        this.addNewTranscription()
      },
    },
  }
</script>

<style scoped lang="scss">
  .list-content {
    overflow: hidden;

    ul {
      list-style: none;
      margin: 0 0 $default-margin 0;
      padding: 0 $default-padding;

      li {
        position: relative;
        margin: 0;
        background-color: #FFFFFF;
        border-top: 1px solid $list-border-color;
        border-right: 1px solid $list-border-color;
        border-left: 1px solid $list-border-color;
        display: flex;
        &:last-child {
          border-bottom: 1px solid $list-border-color;
        }
      }
    }

    &--add-row {
      width: 32px;
      height: 32px;
      margin: 0 auto $default-margin auto;
      display: block;
      cursor: pointer;
    }

    .list-anim-enter-active, .list-anim-leave-active {
      transition: all $default-time $default-easing;
    }

    .list-anim-enter, .list-anim-leave-to {
      opacity: 0;
      margin-left: -100%;
    }
  }
</style>
