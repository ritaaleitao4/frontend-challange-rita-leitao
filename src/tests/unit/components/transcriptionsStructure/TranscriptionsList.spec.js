import Vue from 'vue'
import Vuex from 'vuex'
import { mount, createLocalVue } from '@vue/test-utils'
import TranscriptionsList from '@/components/transcriptionsStructure/TranscriptionsList.vue'

const localVue = createLocalVue()
localVue.use(Vuex)
Vue.use(Vuex)

let actions
let store
let wrapper
let listData = [
  {
    id: 1,
    voice: 'Test Voice',
    text: 'Test description',
  },
  {
    id: 2,
    voice: 'Test Voice 2',
    text: 'Test description 2',
  },
]

describe('Transcriptions list add row Test', () => {
  beforeEach(() => {
    actions = {
      addNewTranscription: jest.fn(() => true),
    }

    store = new Vuex.Store({
      actions,
      modules: {
        transcription: {
          namespaced: true,
          actions,
        },
      },
    })

    wrapper = mount(TranscriptionsList, {
      stubs: ['list-text', 'person-svg', 'CheckboxItem', 'TranscriptionsListRow'],
      store,
      localVue,
      propsData: { listData },
    })
  })

  it('Snapshot test', () => {
    const Constructor = Vue.extend(TranscriptionsList)
    const vm = new Constructor().$mount()
    expect(vm.$el).toMatchSnapshot()
  })

  it('should has a button to add', () => {
    expect(wrapper.contains('.list-content--add-row')).toBe(true)
  })

  it('should call action to add row', () => {
    wrapper.find('.list-content--add-row').trigger('click')
    expect(actions.addNewTranscription).toHaveBeenCalled()
  })

  it('should render a list of items passed as props', () => {
    expect(wrapper.findAll('li').length).toBe(2)
  })

  it('should not render a list of items when an empty array is passed in props', () => {
    listData = []
    wrapper = mount(TranscriptionsList, {
      store,
      localVue,
      propsData: { listData },
    })
    expect(wrapper.findAll('li').length).toBe(0)
  })
})
