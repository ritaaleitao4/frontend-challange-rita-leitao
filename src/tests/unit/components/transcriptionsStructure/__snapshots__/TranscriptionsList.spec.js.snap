// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Transcriptions list add row Test Snapshot test 1`] = `
<div
  class="list-content"
>
  <ul />
   
  <add-row-svg
    class="list-content--add-row"
  />
</div>
`;
